File: frontend/src/api/endpoints/sms.ts
import { apiRequest } from '../client';
import type { QueryResult, SmsQueryRequest } from '../types';

export function querySms(deviceId: string, payload: SmsQueryRequest): Promise<QueryResult> {
  return apiRequest<QueryResult, SmsQueryRequest>(`/devices/${encodeURIComponent(deviceId)}/sms/query`, {
    method: 'POST',
    body: payload,
  });
}

File: app/api/v1/sms.py
from typing import Annotated

from fastapi import APIRouter, Depends, Request

from app.api.deps import CurrentUser, DbSession, require_permission
from app.schemas.sms import SmsQueryRequest
from app.services.command_service import CommandService
from app.utils.responses import success_response

router = APIRouter(prefix="/devices/{device_id}/sms", tags=["sms"])
SmsViewer = Annotated[object, Depends(require_permission("device:sms:query"))]


@router.post("/query")
async def query_sms(
    device_id: str,
    payload: SmsQueryRequest,
    request: Request,
    db: DbSession,
    user: CurrentUser,
    _: SmsViewer,
) -> dict[str, object]:
    data = await CommandService(db).dispatch(device_id, "sms.query", payload.model_dump(), user.id)
    return success_response(request, data)
